xlml/utils/gpu.py

```diff
@@ -285,8 +311,71 @@ def wait_for_resource_creation(operation_name: airflow.XComArg):
           logging.warning(f" - {warning.code}: {warning.message}")
       return True
 
+  @provide_session
+  def retry_all_upstream_tasks(context, session=None):
+    """Retries all upsteam tasks of current task.
+
+    The function searchs for all the upstream tasks and clear those
+    task instances.
+
+    Args:
+      context: Airflow context that provides task info.
+      session: Ariflow session info for clearing tasks.
+    """
+    current_task = context["task"]
+    visited = set()
+    queue = [current_task]
+    # Graph search for all upstream tasks.
+    while queue:
```

Comment:
To be honest, I'm really not comfortable with this sort of graph introspection.  One of the benefits of airflow IMO is that it's very easy to look at the graph view and see what the control flow would be. Tasks modifying each others' states starts to destroy this expectation. It's going to get messy and hard to reason about very quickly. Let's try to find another solution.

Reply:
Right, I will switch to GKE instead of modifying each others task status.